### PR TITLE
feature: allow interaction handlers of any observable type

### LINKED ISSUE #1260: Make Rx-based interactions agnostic about observable type
**Do you want to request a *feature* or report a *bug*?**

Design improvement.

**What is the current behavior?**

Interaction handlers registered via `RegisterHandler(Func<InteractionContext<TInput, TOutput>, IObservable<Unit>> handler)` are required to return `IObservable<Unit>`. It's easy to miss this and return an observable of a different type when calling `RegisterHandler`:

```C#
someInteraction.RegisterHandler(context => someObservableThatIsNotUnit);
```
This will still compile, but resolve to the wrong `RegisterHandler` overload (the one that takes `Action<InteractionContext<TInput, TOutput>>`.

It would be better if `RegisterHandler` took `IObservable<TDontCare>`. The only slight issue with doing so is that every handler received will have to be converted to a signal before it can be stored in the list of handlers. This means that if the observable is already of type `Unit` you will incur a very slight performance hit that you wouldn't otherwise. But I'm only mentioning it for completeness - clearly correctness and falling into a pit of success is far more important.

**What is the expected behavior?**

Just allow any kind of observable to resolve to the correct `RegisterHandler` overload, much the same as any kind of `Task` resolves to the corrent `RegisterHandler` overload.

**What is the motivation / use case for changing the behavior?**

Pit of success, framework design.

**Which versions of ReactiveUI, and which platform / OS are affected by this issue? Did this work in previous versions of ReativeUI? Please also test with the latest stable and snapshot (http://docs.reactiveui.net/en/contributing/snapshot/index.html) versions.**

7

### PR DESCRIPTION
**Pull request by kentcb**
Fixes #1260.

### CODE REVIEW

File: src/ReactiveUI/Interactions.cs [hunk 1]
```diff
@@ -200,14 +200,16 @@ public IDisposable RegisterHandler(Func<InteractionContext<TInput, TOutput>, Tas
         /// <returns>
         /// A disposable which, when disposed, will unregister the handler.
         /// </returns>
-        public IDisposable RegisterHandler(Func<InteractionContext<TInput, TOutput>, IObservable<Unit>> handler)
+        public IDisposable RegisterHandler<TDontCare>(Func<InteractionContext<TInput, TOutput>, IObservable<TDontCare>> handler)
```
Comment (ghuntley) [inline src/ReactiveUI/Interactions.cs]: Should we keep `public IDisposable RegisterHandler(Func<InteractionContext<TInput, TOutput>, IObservable<Unit>> handler)` as an alias method to the new signature but annotate it with `[Obsolete("Please use RegisterHandler<T> instead?")]` as a migration path? I want to stop making hard public API contract changes. Your thoughts?

**Comment by kentcb** [comment]
@ghuntley it's not a breaking change in the sense that if people were accidentally calling the non-observable version because their observable was not `Unit`, this change will actually _help_ them.

**Comment by kentcb** [comment]
@ghuntley not sure which docs you mean? Nothing has changed from a consumer point of view apart from the API being more forgiving of the observable type. But the docs don't talk about that any way.

**Comment by ghuntley** [comment]
Apologies, an assumption on my part that documentation would need updating. If they don't, they don't. I've started thinking about making this a standard question as merging pull requests - does the doc need updating and if so please do it. I'm keen to merge the documentation repo back into the main RxUI repo at some stage this year so documentation gets versioned along side the code.

**Pull request merged**